nixos/modules/virtualisation/openstack-config.nix

```diff
@@ -0,0 +1,72 @@
+{ pkgs, lib, ... }:
+
+with lib;
+
+{
+  imports = [
+    ../profiles/qemu-guest.nix
+    ../profiles/headless.nix
+    # The Openstack Metadata service exposes data on an EC2 API also.
+    ./ec2-data.nix
+    ./amazon-init.nix
+  ];
+
+  config = {
+    fileSystems."/" = {
+      device = "/dev/disk/by-label/nixos";
+      autoResize = true;
+    };
+
+    boot.growPartition = true;
+    boot.kernelParams = [ "console=ttyS0" ];
+    boot.loader.grub.device = "/dev/vda";
+    boot.loader.timeout = 0;
+
+    # Allow root logins
+    services.openssh = {
+      enable = true;
+      permitRootLogin = "prohibit-password";
+      passwordAuthentication = mkDefault false;
+    };
+
+    # Force getting the hostname from Nova.
+    networking.hostName = mkDefault "";
+
+    systemd.services.nova-init = {
+      path = [ pkgs.wget ];
+      description = "Fetch Metadata on startup";
+      wantedBy = [ "multi-user.target" ];
+      before = [ "apply-ec2-data.service" "amazon-init.service"];
+      wants = [ "network-online.target" ];
+      after = [ "network-online.target" ];
+      script =
+        ''
```

Comment:
The following script looks pretty similar to what's in nixos/modules/virtualisation/amazon-image.nix at `boot.initrd.postMountCommands`.

This duplicates the script in nixos/modules/virtualisation/amazon-image.nix, and probably should be merged.

Other suggestion, what about using https://github.com/coreos/coreos-metadata, which already handles changing hostname and ssh keys, and where it should be fairly trivial to make it expose user-data?

Reply:
As discussed, I've moved the script in a shared file.